api/R/insert_new_workflow.R

```diff
@@ -0,0 +1,116 @@
+#' Insert a new workflow into PEcAn database, returning the workflow
+#' as a `data.frame`
+#'
+#' @inheritParams param_query
+#' @param site_id Site ID from `sites` table (numeric)
+#' @param model_id Model ID from `models` table (numeric)
+#' @param start_date Model run start date (character or POSIX)
+#' @param end_date Model run end date (character or POSIX)
+#' @param user_id User ID from `users` table (numeric or NULL (default))
+#' @param hostname Workflow server hostname (character; default =
+#'   option `pecanapi.workflow_hostname`)
+#' @param folder_prefix Output directory prefix (character; default =
+#'   option `pecanapi.workflow_prefix`). Workflow ID will be appended
+#'   to the end with `paste0`
+#' @param params Additional workflow parameters, stored in
+#'   `workflows.params` (character or NULL (default))
+#' @param notes Additional workflow notes, stored in `workflows.notes`
+#'   (character or NULL (default))
+#' @return `data.frame` containing new workflow(s), including all
+#'   columns from `workflows` table.
+#' @author Alexey Shiklomanov
+#' @export
+insert_new_workflow <- function(con,
+                                site_id,
+                                model_id,
+                                start_date,
+                                end_date,
+                                user_id = NULL,
+                                hostname = getOption("pecanapi.workflow_hostname"),
+                                folder_prefix = getOption("pecanapi.workflow_prefix"),
+                                params = NULL,
+                                notes = NULL) {
+  if (is.null(notes)) notes <- ""
+  if (is.null(params)) params <- ""
+  stopifnot(
+    # Must be scalar 
+    length(folder_prefix) == 1,
+    length(user_id) <= 1,
+    length(con) == 1,
+    # Must be RPostgres connection for prepared queries 
+    class(con) == "PqConnection"
+  )
+  lens <- lengths(list(site_id, model_id, start_date, end_date))
+  n_workflow <- max(lens)
+  if (!all(lens == 1 | lens == n_workflow)) {
+    stop(
+      "All inputs must be either the same length or length 1. ",
+      "You provided the following: ",
+      paste(sprintf("%s (%s)", c("site_id", "model_id", "start_date", "end_date"), lens),
+            collapse = ", ")
+    )
+  }
+  id <- bit64::integer64()
+  for (i in seq_len(n_workflow)) {
+    id[i] <- get_next_workflow_id(con)[[1]]
+  }
+  stopifnot(length(id) >= 1)
+  folder <- paste0(folder_prefix, id)
+  query_string <- paste(
+    "INSERT INTO workflows",
+    "(id, site_id, model_id, folder,",
```

Comment:
I don't understand why you're taking 2 steps to get the next workflow ID then inserting that ID in directly rather than just inserting without the ID and letting the database assign the next ID? The latter is both simpler and safer (because, in theory, another workflow could be inserting a record between your ID query and your INSERT)

Reply:
The reason I did this is that the workflow `folder` depends on the ID, and there's no way to construct it without having the ID in advance. An alternative is to run two queries -- first, `INSERT` the workflow (automatically creating the ID), then `UPDATE` those rows with the folder, but that seems equally clunky.

There is no safety concern here for two reasons. First, because the act of retrieving an ID via `nextval` also automatically augments the internal workflow ID counter in PostgreSQL (see [`nextval` documentation](https://www.postgresql.org/docs/9.6/functions-sequence.html)). Second, because the workflow table has a constraint that the `id` column is unique, meaning that it will error if I try to create two workflows with the same ID.